File: skrelief/Relief.jl
module Relief
export relief

using StatsBase
include("./utils/square_to_vec.jl")


"""
    relief(data::Array{<:AbstractFloat,2}, target::Array{<:Integer,1}, m::Signed=-1, 
                dist_func::Function=(e1, e2) -> sum(abs.(e1 .- e2), dims=2), f_type::String="continuous")::Array{Float64,1}

Compute feature weights using Relief algorithm. The f_type argument specifies whether the features are continuous or discrete 
and can either have the value of "continuous" or "discrete".

---
# Reference:
- Kenji Kira and Larry A. Rendell. The feature selection problem: Tra-
ditional methods and a new algorithm. In Proceedings of the Tenth
National Conference on Artificial Intelligence, AAAI’92, pages 129–134.
AAAI Press, 1992.
"""
function relief(data::Array{<:Number,2}, target::Array{<:Integer,1}, m::Signed=-1, 
                dist_func::Function=(e1, e2) -> sum(abs.(e1 .- e2), dims=2); f_type::String="continuous")::Array{Float64,1}

    # Initialize feature weights vector.
    weights = zeros(Float64, size(data, 2))

    # Compute vectors of maximum and minimum feature values.
    max_f_vals = vec(maximum(data, dims=1))
    min_f_vals = vec(minimum(data, dims=1))
    
    # Sample m examples without replacement. If m has signal value -1, use all examples.
    sample_idxs = StatsBase.sample(1:size(data, 1), if (m==-1) size(data,1) else m end, replace=false)
    if (m == -1) m = size(data, 1) end # If m has signal value -1, set m to total number of examples.

    # Compute pairwise distances between samples (vector form).
    # Note that the number of elements in the distance vector is {n \choose 2} = n!/(2!*(n-2)!) = n*(n-1)/2.
    # Add additional element with 0.
    dists = Array{Float64}(undef, Int64((size(data, 1)^2 - size(data, 1))/2 + 1))
    dists[1] = 0  # Set first value of distances vector to 0 - accessed when i == j in square form indices.

    # Construct pairwise distances vector using vectorized distance function.
    top_ptr = 2
    @inbounds for idx = 1:size(data,1)-1
        upper_lim = top_ptr + size(data, 1) - idx - 1
        dists[top_ptr:upper_lim] = dist_func(data[idx:idx, :], data[idx+1:end, :])
        top_ptr = upper_lim + 1
    end

    # Go over sampled indices.
    @inbounds for idx = sample_idxs

        # Row and column indices for querying pairwise distance vector.
        row_idxs = repeat([idx - 1], size(data, 1))
        col_idxs = collect(0:size(data, 1)-1)
        
        # Get indices of neighbours with same class in distances vector and find nearest hit.
        neigh_idx_hit = Int64.(square_to_vec(row_idxs[target .== target[idx]], col_idxs[target .== target[idx]], size(data, 1))) .+ 2
        idx_nearest_hit = partialsortperm(dists[neigh_idx_hit], 1:2)[2:end]
        nearest_hit = vec(data[target .== target[idx], :][idx_nearest_hit, :])

        # Get indices of neighbours with different class in distances vector and find nearest miss.
        neigh_idx_miss = Int64.(square_to_vec(row_idxs[target .!= target[idx]], col_idxs[target .!= target[idx]], size(data, 1))) .+ 2
        idx_nearest_miss = argmin(dists[neigh_idx_miss])
        nearest_miss = vec(data[target .!= target[idx], :][idx_nearest_miss, :])
        
        ### Weights Update ###
       
        if f_type == "continuous"
            # If features continuous.
            weights = weights .- (abs.(data[idx, :] .- nearest_hit)./(max_f_vals .- min_f_vals .+ eps(Float64)))./m .+
                (abs.(data[idx, :] .- nearest_miss)./(max_f_vals .- min_f_vals .+ eps(Float64)))./m 
        elseif f_type == "discrete"
            # If features discrete.
            weights = weights .- Int64.(data[idx, :] .!= nearest_hit)./m .+ Int64.(data[idx, :] .!= nearest_miss)./m
        else
            throw(DomainError(f_type, "f_type can only be equal to \"continuous\" or \"discrete\"."))
        end

        ######################

    end

    # Return computed feature weights.
    return weights
end 

end

File: skrelief/relief.py
import numpy as np
from scipy.stats import rankdata
from sklearn.base import BaseEstimator, TransformerMixin
import os

from julia import Julia
jl = Julia(compiled_modules=False)
script_path = os.path.abspath(__file__)
jl.eval('push!(LOAD_PATH, "' + script_path[:script_path.rfind('/')] + '/")')

from julia import Relief as Relief_jl


class Relief(BaseEstimator, TransformerMixin):
    """sklearn compatible implementation of the Relief algorithm.

    Reference:
        Kenji Kira and Larry A. Rendell. The feature selection problem: Tra-
        ditional methods and a new algorithm. In Proceedings of the Tenth
        National Conference on Artificial Intelligence, AAAI’92, pages 129–134.
        AAAI Press, 1992.

    Args:
        n_features_to_select (int): number of features to select from dataset.
        m (int): training data sample size.
        dist_func (function): function used to measure similarities between samples.

    Attributes:
        n_features_to_select (int): number of features to select from dataset.
        m (int): training data sample size.
        dist_func (function): function used to measure similarities between samples.
        _relief (function): function implementing Relief algorithm written in Julia programming language.
    """

    def __init__(self, n_features_to_select=10, m=-1, dist_func=lambda x1, x2: np.sum(np.abs(x1 - x2), 1)):
        self.n_features_to_select = n_features_to_select
        self.m = m
        self.dist_func = dist_func
        self._relief = Relief_jl.relief


    def fit(self, data, target):
        """
        Rank features using Relief feature selection algorithm

        Args:
            data (numpy.ndarray): matrix of data samples
            target (numpy.ndarray): vector of target values of samples

        Returns:
            (object): reference to self
        """

        # Compute feature weights and rank.
        self.weights = self._relief(data, target, self.m, self.dist_func)
        self.rank = rankdata(-self.weights, method='ordinal')

        # Return reference to self
        return self


    def transform(self, data):
        """
        Perform feature selection using computed feature ranks.

        Args:
            data (numpy.ndarray): matrix of data samples on which to perform feature selection.

        Returns:
            (numpy.ndarray): result of performing feature selection.
        """

        # select n_features_to_select best features and return selected features.
        msk = self.rank <= self.n_features_to_select  # Compute mask.
        return data[:, msk]  # Perform feature selection.

